doc_gen.php

```diff
@@ -0,0 +1,164 @@
+        $string,
+        $version,
+        $filename,
+        $startLine
+    );
+}
+
+function normaliseFilename($filename): string
+{
+    $lastPos = strrpos($filename, 'psl/src');
+    if ($lastPos !== false) {
+        $filename = substr($filename, $lastPos + strlen('psl/src/'));
+    }
+    return $filename;
+}
+
+function generateFunctionLine($sha, $function): string
+{
+    $rf = new ReflectionFunction($function);
+
+    $filename = $rf->getFileName();
+    $filename = normaliseFilename($filename);
+
+    $link = generateLink($sha, $filename, $rf->getStartLine());
+    $rf = new ReflectionFunction($function);
+
+    $line = generateNamespaceLine($rf->getNamespaceName());
+
+    $functionWithoutNamespace = str_replace($rf->getNamespaceName(), '', $function);
+    $functionWithoutNamespace = ltrim($functionWithoutNamespace, '\\');
+
+    $line .= sprintf(
+        "* <a href='%s'>%s</a>  ", //double space for new line
+        $link,
+        $functionWithoutNamespace
+    );
+
+    return $line;
+}
+
+function generateInterfaceLine($sha, $interface): string
+{
+    $rc = new ReflectionClass($interface);
+
+    $filename = $rc->getFileName();
+    $filename = normaliseFilename($filename);
+
+    $link = generateLink($sha, $filename, $rc->getStartLine());
+
+    $line = generateNamespaceLine($rc->getNamespaceName());
+
+    $functionWithoutNamespace = str_replace($rc->getNamespaceName(), '', $interface);
+    $functionWithoutNamespace = ltrim($functionWithoutNamespace, '\\');
+
+    $line .= sprintf(
+        "* <a href='%s'>%s</a>  ", //double space for new line
+        $link,
+        $functionWithoutNamespace
+    );
+
+    return $line;
+}
+
+function generateDocList(): string
+{
+    $version = '1.5.x';
+
+    require_once __DIR__ . "/vendor/autoload.php";
+
+    $contents = [];
+    $contents[] = '## Functions';
+    $contents[] = '';
+    foreach (\Psl\Internal\Loader::FUNCTIONS as $function) {
+        $contents[] = generateFunctionLine($version, $function);
+    }
+
+    $contents[] = '';
+    $contents[] = '## Interfaces';
+    $contents[] = '';
+    foreach (\Psl\Internal\Loader::INTERFACES as $interface) {
+        $contents[] = generateInterfaceLine($version, $interface);
+    }
+
+    $contents[] = '';
+    $contents[] = '## Classes';
+    $contents[] = '';
+    foreach (\Psl\Internal\Loader::CLASSES as $interface) {
+        $contents[] = generateInterfaceLine($version, $interface);
+    }
+
+    $contents[] = '---';
+    $contents[] = 'This markdown file was generated from doc_gen.php. Any edits to it will likely be lost.';
+
+    return implode("\n", $contents);
+}
+
+
+$contents = <<< MD
+# Documentation
+
+This doc contains a list of the functions, interfaces and classes this library provides.
+
+Please click through to read the docblock comment details for each of them.
+
+MD;
+
+$contents .= generateDocList();
+
+function checkDocsUpToDate($docFilename, $contents)
+{
+    $newLines = explode("\n", $contents);
```

Comment:
You can also use `Psl\Shell\execute()` with `git diff --exit-code $docFilename`.
This gives you a nice diff of the changed lines and returns the correct exit code.